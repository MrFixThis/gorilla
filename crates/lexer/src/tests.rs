use std::collections::VecDeque;

use super::*;

macro_rules! snapshot {
    ( $func:ident, $path:expr ) => {
        #[test]
        fn $func() {
            let src = include_str!($path);
            let mut conf = ::insta::Settings::clone_current();
            conf.set_snapshot_path("../testdata/output/");
            conf.bind(|| ::insta::assert_snapshot!(prepare_snapshot(src)));
        }
    };
}

fn prepare_snapshot(src: &str) -> String {
    let lexer = Lexer::new(src);
    let mut toks: VecDeque<Token> = VecDeque::new();

    loop {
        let tok = lexer.next_tok();
        if tok.kind == Eof {
            break toks.push_back(tok);
        }
        toks.push_back(tok);
    }

    let mut out = String::new();
    for (row, line) in src.lines().enumerate() {
        out += line;
        out += "\n";

        while let Some(tok) = toks.pop_front() {
            if tok.span.start_row != row && tok.kind != Eof {
                toks.push_front(tok);
                break;
            }

            out += &" ".repeat(tok.span.start_col);
            out += &"^".repeat((tok.span.end_col - tok.span.start_col) + 1);
            out += &format!(" {tok:?}\n");
        }
    }

    out
}

snapshot!(test_code_snipet_tokenization, "../testdata/snapshots/code_snipet.gl");
snapshot!(test_comment_tokenization, "../testdata/snapshots/comment.gl");
snapshot!(test_container_types_tokenization, "../testdata/snapshots/container_type.gl");
snapshot!(test_literal_error_tokenization, "../testdata/snapshots/literal_error.gl");
snapshot!(test_function_tokenization, "../testdata/snapshots/function.gl");
snapshot!(test_identifier_tokenization, "../testdata/snapshots/identifier.gl");
snapshot!(test_literal_tokenization, "../testdata/snapshots/literal.gl");
