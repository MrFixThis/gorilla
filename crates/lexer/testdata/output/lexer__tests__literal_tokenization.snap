---
source: crates/lexer/src/tests.rs
expression: prepare_snapshot(src)
---
var my_str = "Rust is my pather";
^^^ Keyword(var) -> [0 - 0, 0 - 2]
    ^^^^^^ Ident(my_str) -> [0 - 4, 0 - 9]
           ^ Eq -> [0 - 11, 0 - 11]
             ^^^^^^^^^^^^^^^^^^ Literal(Str, "Rust is my pather") -> [0 - 13, 0 - 30]
                                ^ Semicolon -> [0 - 32, 0 - 32]
var my_bool = true;
^^^ Keyword(var) -> [1 - 0, 1 - 2]
    ^^^^^^^ Ident(my_bool) -> [1 - 4, 1 - 10]
            ^ Eq -> [1 - 12, 1 - 12]
              ^^^^ Keyword(true) -> [1 - 14, 1 - 17]
                  ^ Semicolon -> [1 - 18, 1 - 18]
var my_int = 13;
^^^ Keyword(var) -> [2 - 0, 2 - 2]
    ^^^^^^ Ident(my_int) -> [2 - 4, 2 - 9]
           ^ Eq -> [2 - 11, 2 - 11]
             ^^ Literal(Int, 13) -> [2 - 13, 2 - 14]
               ^ Semicolon -> [2 - 15, 2 - 15]
var my_float = 5.0;
^^^ Keyword(var) -> [3 - 0, 3 - 2]
    ^^^^^^^^ Ident(my_float) -> [3 - 4, 3 - 11]
             ^ Eq -> [3 - 13, 3 - 13]
               ^^^ Literal(Float, 5.0) -> [3 - 15, 3 - 17]
                  ^ Semicolon -> [3 - 18, 3 - 18]
^ Eof -> [0 - 0, 0 - 0]

